[SDK52] [expo-updates] android build failure: Inconsistent JVM-target compatibility detected for tasks 'compileDebugJavaWithJavac' (17) and 'kspDebugKotlin' (21).

### Minimal reproducible example

https://github.com/rranjan14/expo-update-test

### What platform(s) does this occur on?

Android

### Where did you reproduce the issue?

in a development build

### Summary

ksp version is too old for the kotlin version being used. Happening with expo-updates.
<img width="1259" alt="Screenshot 2024-11-20 at 16 19 37" src="https://github.com/user-attachments/assets/fa69a5e7-18b9-49fb-9282-5a09ef72ba80">

JVM version I am using is 17


### Environment

```text
expo-env-info 1.2.1 environment info:
    System:
      OS: macOS 15.1.1
      Shell: 3.7.1 - /opt/homebrew/bin/fish
    Binaries:
      Node: 20.13.1 - ~/.nvm/versions/node/v20.13.1/bin/node
      Yarn: 1.22.22 - ~/.nvm/versions/node/v20.13.1/bin/yarn
      npm: 10.9.0 - /opt/homebrew/bin/npm
      Watchman: 2024.11.18.00 - /opt/homebrew/bin/watchman
    SDKs:
      iOS SDK:
        Platforms: DriverKit 24.1, iOS 18.1, macOS 15.1, tvOS 18.1, visionOS 2.1, watchOS 11.1
      Android SDK:
        API Levels: 33, 34, 35
        Build Tools: 30.0.2, 30.0.3, 33.0.0, 34.0.0, 35.0.0, 35.0.0
        System Images: android-UpsideDownCakePrivacySandbox | Google Play ARM 64 v8a
    IDEs:
      Android Studio: 2024.2 AI-242.23339.11.2421.12550806
      Xcode: 16.1/16B40 - /usr/bin/xcodebuild
    npmPackages:
      expo: ~52.0.8 => 52.0.8
      react: 18.3.1 => 18.3.1
      react-native: 0.76.2 => 0.76.2
    npmGlobalPackages:
      eas-cli: 13.1.1
    Expo Workflow: managed
```


### Expo Doctor Diagnostics

```text
Enabled experimental React Native Directory checks. Unset the EXPO_DOCTOR_ENABLE_DIRECTORY_CHECK environment variable to disable this check.
✔ Check package.json for common issues
✔ Check Expo config for common issues
✔ Check native tooling versions
✔ Check if the project meets version requirements for submission to app stores
✔ Check for common project setup issues
✔ Check dependencies for packages that should not be installed directly
✔ Check for app config fields that may not be synced in a non-CNG project
✔ Check for issues with Metro config
✔ Check npm/ yarn versions
✔ Validate packages against React Native Directory package metadata
✔ Check Expo config (app.json/ app.config.js) schema
✔ Check that native modules do not use incompatible support packages
✔ Check for legacy global CLI installed locally
✔ Check that native modules use compatible support package versions for installed Expo SDK
✔ Check that packages match versions required by installed Expo SDK

Didn't find any issues with the project!
```

I am getting the same error also (although I suspect mine was related to moving from a 2019 MacBook Pro to a new M4 Max MacBook Pro (and I have had to reinstall several coding related packages to move from `x64` to `arm64` packages.  Android Studio, Android SDKs, Java JRE and JDK have been the problem children.

Output from my latest Android development build:

```
[RUN_GRADLEW] FAILURE: Build failed with an exception.
[RUN_GRADLEW] * What went wrong:
[RUN_GRADLEW] Execution failed for task ':expo-updates:kspDebugKotlin'.
[RUN_GRADLEW] > Inconsistent JVM-target compatibility detected for tasks 'compileDebugJavaWithJavac' (17) 
and 'kspDebugKotlin' (21).
[RUN_GRADLEW]   
[RUN_GRADLEW]   Consider using JVM Toolchain: https://kotl.in/gradle/jvm/toolchain
[RUN_GRADLEW]   Learn more about JVM-target validation: https://kotl.in/gradle/jvm/target-validation
```

Same problem for me: I have Java 21 installed in my machine and my project used to build fine with Expo 51, but as soon as I upgraded to Expo 52 I started to get the error below:

```
> Task :expo-updates:kspDebugKotlin FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':expo-updates:kspDebugKotlin'.
> Inconsistent JVM-target compatibility detected for tasks 'compileDebugJavaWithJavac' (17) and 'kspDebugKotlin' (21).

  Consider using JVM Toolchain: https://kotl.in/gradle/jvm/toolchain
  Learn more about JVM-target validation: https://kotl.in/gradle/jvm/target-validation
```

The problem seems to be coming from `expo-updates`.

--- a/packages/expo-modules-core/CHANGELOG.md
+++ b/packages/expo-modules-core/CHANGELOG.md
@@ -12,6 +12,8 @@
 
 ### 💡 Others
 
+- [Android] Introduced `applyKspJvmToolchain()` gradle helper to enforce JVM Toolchain version for KSP. ([#33148](https://github.com/expo/expo/pull/33148) by [@kudo](https://github.com/kudo))
+
 ## 2.0.4 — 2024-11-19
 
 ### 🎉 New features

--- a/packages/expo-modules-core/android/ExpoModulesCorePlugin.gradle
+++ b/packages/expo-modules-core/android/ExpoModulesCorePlugin.gradle
@@ -51,6 +51,13 @@ ext.applyKotlinExpoModulesCorePlugin = {
   apply plugin: KotlinExpoModulesCorePlugin
 }
 
+// Apply JVM Toolchain version for KSP
+ext.applyKspJvmToolchain = {
+  project.ksp {
+    kotlin.jvmToolchain(17)
+  }
+}
+
 // Setup build options that are common for all modules
 ext.useDefaultAndroidSdkVersions = {
   project.android {

--- a/packages/expo-updates/CHANGELOG.md
+++ b/packages/expo-updates/CHANGELOG.md
@@ -8,6 +8,8 @@
 
 ### 🐛 Bug fixes
 
+- Fixed `Inconsistent JVM-target compatibility` error when building with JDK 21. ([#33148](https://github.com/expo/expo/pull/33148) by [@kudo](https://github.com/kudo))
+
 ### 💡 Others
 
 - Introduced `ReactNativeFeatureFlags` compat to fix React Native 0.77 breaking changes. ([#33077](https://github.com/expo/expo/pull/33077) by [@kudo](https://github.com/kudo))

--- a/packages/expo-updates/android/build.gradle
+++ b/packages/expo-updates/android/build.gradle
@@ -21,6 +21,7 @@ version = '0.26.8'
 def expoModulesCorePlugin = new File(project(":expo-modules-core").projectDir.absolutePath, "ExpoModulesCorePlugin.gradle")
 apply from: expoModulesCorePlugin
 applyKotlinExpoModulesCorePlugin()
+applyKspJvmToolchain()
 useCoreDependencies()
 useDefaultAndroidSdkVersions()
 useExpoPublishing()

